### ISSUE TITLE
Complex dot product performance of CuArrays and of StructArrays of CuArrays

### ISSUE THREAD

**Issue by coezmaden** [JuliaGPU/CUDA.jl#667]
**Describe the bug**

1) Dot product of StructArrays of CuArrays is considerably slower than its plain CuArray counterpart.
2) Dot product of a complex vector with a real vector results in scalar indexing

Tested are several constellations of input parameters best described in the tables provided below (Arrays size of 10000, italic = scalar indexing). 

| CuArray | real · real | cplx · cplx | real · cplx | dot_parts(cplx, real) | dot_padded(cplx, real) |
| -------- | ----------- | -----------| -----------| ----------------------|--------------------------|
| btime w/ sync | 90.900 μs | 91.000 μs | _362.500 ms_ | 135.900 μs | 90.299 μs |

| StructArray | real · real | cplx · cplx | real · cplx | dot_struct_cplx_cplx | dot_struct_cplx_real| 
| -------------| --------- | ----------- |  ------------|----------------------|-----------| 
| btime w/ sync | N/A   | _750.751 ms_ |  _567.323 ms_   | 228.099 μs | 118.199 μs |

**To reproduce**
```julia
using CUDA, StructArrays, LinearAlgebra, BenchmarkTools

# Initialize the vectors
N = 10000
real_vector = CUDA.ones(Float32, N)
cplx_vector = CUDA.ones(ComplexF32, N)
CUDA.allowscalar(true) # Allow for scalar indexing for proper intitialization of StructArray
cplx_struct = StructArray(cplx_vector) 
CUDA.allowscalar(false) # Turn it off once we're done

## Custom dot product functions

# Perform dot product by taking the complex vector apart
function dot_parts(cplx_vector::CuArray, real_vector::CuArray)
    complex.(dot(real(cplx_vector), real_vector), dot(imag(cplx_vector), real_vector))
end

# Transform a real vector into a complex one by padding its imaginary part with zeros in order to invoke dotc
function dot_padded(cplx_vector::CuArray, real_vector::CuArray)
    dot(cplx_vector, complex.(real_vector, CUDA.zeros(length(real_vector))))
end

# Perform the complex dot product by dividing the StructArray into real and imaginary parts
function dot_struct_cplx_cplx(cplx_struct_1::StructArray, cplx_struct_2::StructArray)
    complex.(
        dot(cplx_struct_1.re, cplx_struct_2.re) - dot(cplx_struct_1.im, cplx_struct_2.im),
        dot(cplx_struct_1.re, cplx_struct_2.im) + dot(cplx_struct_1.im, cplx_struct_2.re)
    )
end

# Perform an ordinary dot product by dividing the StructArray into real and imaginary parts
function dot_struct_cplx_real(cplx_struct::StructArray, real_vector::CuArray)
    complex.(dot(cplx_struct.re, real_vector), dot(cplx_struct.im, real_vector))
end

## Benchmarks

@btime CUDA.@sync dot($real_vector,$real_vector)
@btime CUDA.@sync dot($cplx_vector,$cplx_vector)
@btime CUDA.@sync dot($cplx_vector,$real_vector) # scalar indexing
@btime CUDA.@sync dot_parts($cplx_vector,$real_vector)
@btime CUDA.@sync dot_padded($cplx_vector,$real_vector)
@btime CUDA.@sync dot($cplx_struct,$cplx_struct) # scalar indexing
@btime CUDA.@sync dot_struct_cplx_cplx($cplx_struct,$cplx_struct)
@btime CUDA.@sync dot($cplx_struct,$real_vector) # scalar indexing
@btime CUDA.@sync dot_struct_cplx_real($cplx_struct,$real_vector)


```

<details><summary>Manifest.toml</summary>
<p>

```
[[CUDA]]
deps = ["AbstractFFTs", "Adapt", "BFloat16s", "CEnum", "CompilerSupportLibraries_jll", "DataStructures", "ExprTools", "GPUArrays", "GPUCompiler", "LLVM", "Libdl", "LinearAlgebra", "Logging", "MacroTools", "NNlib", "Pkg", "Printf", "Random", "Reexport", "Requires", "SparseArrays", "Statistics", "TimerOutputs"]
git-tree-sha1 = "39f6f584bec264ace76f924d1c8637c85617697e"
uuid = "052768ef-5323-5732-b1bb-66c8b64840ba"
version = "2.4.0"

[[GPUArrays]]
deps = ["AbstractFFTs", "Adapt", "LinearAlgebra", "Printf", "Random", "Serialization"]
git-tree-sha1 = "f99a25fe0313121f2f9627002734c7d63b4dd3bd"
uuid = "0c68f7d7-f131-5f86-a1c3-88cf8149b2d7"
version = "6.2.0"

[[LLVM]]
deps = ["CEnum", "Libdl", "Printf", "Unicode"]
git-tree-sha1 = "d0d99629d6ae4a3e211ae83d8870907bd842c811"
uuid = "929cbde3-209d-540e-8aea-75f648917ca0"
version = "3.5.2"


```

</p>
</details>

**Expected behavior**
1) Similar or even better performance of StructArrays of CuArrays regarding the complex dot product

2) Support for real · cplx without scalar indexing.

**Version info**

Details on Julia:

```
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9600K CPU @ 3.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = "C:\Program Files\Microsoft VS Code\Code.exe"
  JULIA_NUM_THREADS =
```

Details on CUDA:

```
CUDA toolkit 11.1.1, artifact installation
CUDA driver 11.1.0
NVIDIA driver 456.71.0

Libraries:
- CUBLAS: 11.3.0
- CURAND: 10.2.2
- CUFFT: 10.3.0
- CUSOLVER: 11.0.1
- CUSPARSE: 11.3.0
- CUPTI: 14.0.0
- NVML: 11.0.0+451.22
- CUDNN: 8.0.4 (for CUDA 11.1.0)
- CUTENSOR: 1.2.1 (for CUDA 11.1.0)

Toolchain:
- Julia: 1.5.3
- LLVM: 9.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75

1 device:
  0: GeForce GTX 1050 Ti (sm_61, 3.189 GiB / 4.000 GiB available)
```

**Additional context**

Noticed performance differences/issues while trying to implement real · cplx both for CuArrays and StructArrays of CuArrays. I believe it would be nice to provide off the shelve support for this kind of operation.

**Comment by maleadt** [comment]
>  I believe it would be nice to provide off the shelve support for this kind of operation.

CUDA.jl can't possible integrate with every package out there. This would be better suited as an enhancement to StructArrays.jl, possibly gated by Requires.jl.

**Comment by coezmaden** [comment]
Thanks for the quick reply @maleadt !

> CUDA.jl can't possible integrate with every package out there.

Probably expressed myself poorly there. Didn't mean to ask for that. 
Implied an efficient way to deal with ```dot(::CuArray{ComplexF32},::CuArray{Float32})``` for now.

Have you got an idea why the execution times differ for StructArrays? Is it better if repost this in their issue tracker?

**Comment by coezmaden** [comment]
Reduced the problem to pure CUDA.jl. This issue was poorly worded and heavily realiant on StructArrays.jl. New issue: #668 Edit: spelling
